### PR TITLE
feat: added handling for failed step name in CI messages

### PR DESCRIPTION
**Pull request by subhashish-devtron**
# Description

Fixes #
_added handling for failed step name in CI messages, sending failure notification through CI complete topic instead of workflow update topic to be able to pass details from CI runner for run failures_
## How Has This Been Tested?
Tested manually for success and failed cases on events to notifier with the new data
 

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [x] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/router/pubsub/CiEventHandler.go [hunk 1]
```diff
@@ -92,19 +93,29 @@ func (impl *CiEventHandlerImpl) Subscribe() error {
...
 
-		resp, err := impl.webhookService.HandleCiSuccessEvent(ciCompleteEvent.PipelineId, req)
-		if err != nil {
-			impl.logger.Error(err)
-			return
+		if ciCompleteEvent.FailedStepName != "" {
+			req.FailedStepName = ciCompleteEvent.FailedStepName
+			err := impl.webhookService.HandleCiStepFailedEvent(ciCompleteEvent.PipelineId, req)
+			if err != nil {
+				impl.logger.Error(err)
```
Comment (kripanshdevtron) [inline api/router/pubsub/CiEventHandler.go:106]: log error in better way.

File: pkg/pipeline/CiHandler.go [hunk 2]
```diff
@@ -812,7 +812,12 @@ func (impl *CiHandlerImpl) UpdateWorkflow(workflowStatus v1alpha1.WorkflowStatus
 		}
 		if string(v1alpha1.NodeError) == savedWorkflow.Status || string(v1alpha1.NodeFailed) == savedWorkflow.Status {
 			impl.Logger.Warnw("ci failed for workflow: ", "wfId", savedWorkflow.Id)
-			go impl.WriteCIFailEvent(savedWorkflow, ciWorkflowConfig.CiImage)
+
+			if savedWorkflow.Message != "Error (exit code 2)" {
```
Comment (kripanshdevtron) [inline pkg/pipeline/CiHandler.go:816]: extract error code and check error code using constants

File: pkg/pipeline/WebhookService.go [hunk 3]
```diff
@@ -115,6 +117,24 @@ func (impl WebhookServiceImpl) AuthenticateExternalCiWebhook(apiKey string) (int
 	return id, nil
 }
 
+func (impl WebhookServiceImpl) HandleCiStepFailedEvent(ciPipelineId int, request *CiArtifactWebhookRequest) (err error) {
+
+	savedWorkflow, err := impl.ciWorkflowRepository.FindById(*request.WorkflowId)
+	if err != nil {
+		impl.logger.Errorw("cannot get saved wf", "err", err)
```
Comment (kripanshdevtron) [inline pkg/pipeline/WebhookService.go:124]: log metadata like workflowId

File: pkg/pipeline/WebhookService.go [hunk 4]
```diff
@@ -115,6 +117,24 @@ func (impl WebhookServiceImpl) AuthenticateExternalCiWebhook(apiKey string) (int
...
+		return err
+	}
+
+	pipeline, err := impl.ciPipelineRepository.FindByCiAndAppDetailsById(ciPipelineId)
+	if err != nil {
+		impl.logger.Errorw("unable to find pipeline", "name", request.PipelineName, "err", err)
+		return err
+	}
+
+	impl.WriteCIStepFailedEvent(pipeline, request, savedWorkflow)
```
[the remark this answers is not in the document]
Comment (subhashish-devtron) [reply pkg/pipeline/WebhookService.go:134]: made async

File: api/router/pubsub/CiEventHandler.go [hunk 5]
```diff
@@ -92,19 +93,29 @@ func (impl *CiEventHandlerImpl) Subscribe() error {
...
 
-		resp, err := impl.webhookService.HandleCiSuccessEvent(ciCompleteEvent.PipelineId, req)
-		if err != nil {
-			impl.logger.Error(err)
-			return
+		if ciCompleteEvent.FailureReason != "" {
+			req.FailureReason = ciCompleteEvent.FailureReason
+			err := impl.webhookService.HandleCiStepFailedEvent(ciCompleteEvent.PipelineId, req)
+			if err != nil {
+				impl.logger.Error("Error while sending event for CI failure", "error: ", err)
```
Comment (gireesh-devtron) [inline api/router/pubsub/CiEventHandler.go:106]: log pipelineId and req payload in the log.

File: api/router/pubsub/CiEventHandler.go [hunk 6]
```diff
@@ -92,19 +93,29 @@ func (impl *CiEventHandlerImpl) Subscribe() error {
...
+				return
+			}
+		} else {
+			util.TriggerCIMetrics(ciCompleteEvent.Metrics, impl.ciEventConfig.ExposeCiMetrics, ciCompleteEvent.PipelineName, ciCompleteEvent.AppName)
+
+			resp, err := impl.webhookService.HandleCiSuccessEvent(ciCompleteEvent.PipelineId, req)
+			if err != nil {
+				impl.logger.Error(err)
```
Comment (gireesh-devtron) [inline api/router/pubsub/CiEventHandler.go:114]: same here

File: pkg/pipeline/WebhookService.go [hunk 7]
```diff
@@ -286,6 +307,21 @@ func (impl WebhookServiceImpl) HandleExternalCiWebhook(externalCiId int, request
...
+	material := &client.MaterialTriggerInfo{}
+	material.GitTriggers = ciWorkflow.GitTriggers
+	event.CiWorkflowRunnerId = ciWorkflow.Id
+	event.UserId = int(ciWorkflow.TriggeredBy)
+	event = impl.eventFactory.BuildExtraCIData(event, material, request.Image)
+	event.CiArtifactId = 0
+	event.Payload.FailureReason = request.FailureReason
+	_, evtErr := impl.eventClient.WriteNotificationEvent(event)
+	if evtErr != nil {
+		impl.logger.Errorw("error in writing event", "err", evtErr)
```
Comment (gireesh-devtron) [inline pkg/pipeline/WebhookService.go:321]: log some event information for bebugging

**Review by gireesh-devtron — APPROVED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Review by gireesh-devtron — APPROVED** [review]

**Pull request merged**